crates/unleash-edge-enterprise/src/lib.rs

```diff
@@ -1,42 +1,47 @@
 use std::{pin::Pin, sync::Arc};
-use tokio::sync::oneshot;
-use tracing::{debug, error, info};
+use tokio::sync::watch::Sender;
+use tracing::{debug, info, warn};
 use unleash_edge_http_client::UnleashClient;
-use unleash_edge_types::{errors::EdgeError, tokens::EdgeToken};
+use unleash_edge_types::{RefreshState, errors::EdgeError, tokens::EdgeToken};
 
 pub async fn send_heartbeat(
     unleash_client: Arc<UnleashClient>,
     token: EdgeToken,
-    shutdown_tx: Arc<tokio::sync::Mutex<Option<oneshot::Sender<()>>>>,
+    refresh_state_tx: Sender<RefreshState>,
 ) {
     match unleash_client.send_heartbeat(&token).await {
+        Err(EdgeError::ExpiredLicense(e)) => {
+            warn!("License is expired according to upstream: {}", e);
+        }
         Err(EdgeError::InvalidLicense(e)) => {
-            error!(
-                "License was invalidated by upstream: {}. Shutting down Edge.",
-                e
-            );
-            shutdown_tx.lock().await.take().map(|tx| tx.send(()));
+            warn!("License is invalid according to upstream: {}", e);
+            let _ = refresh_state_tx.send(RefreshState::Paused);
         }
         Err(e) => {
             info!("Unexpected error sending heartbeat: {}", e);
         }
         Ok(_) => {
             debug!("Successfully sent heartbeat");
+            let _ = refresh_state_tx.send(RefreshState::Running);
         }
     }
 }
 
 pub fn create_enterprise_heartbeat_task(
     unleash_client: Arc<UnleashClient>,
     token: EdgeToken,
-    shutdown_tx: oneshot::Sender<()>,
+    refresh_state_tx: Sender<RefreshState>,
 ) -> Pin<Box<dyn Future<Output = ()> + Send>> {
     Box::pin(async move {
         let sleep_duration = tokio::time::Duration::from_secs(90);
-        let shutdown_arc = Arc::new(tokio::sync::Mutex::new(Some(shutdown_tx)));
         loop {
             tokio::time::sleep(sleep_duration).await;
-            send_heartbeat(unleash_client.clone(), token.clone(), shutdown_arc.clone()).await;
+            send_heartbeat(
+                unleash_client.clone(),
+                token.clone(),
+                refresh_state_tx.clone(),
```

Comment:
We don't need to clone here, we can just borrow. It's not an issue here but passing borrows when you can is a good habit to get into